### PR TITLE
feat: implement NEP372 gas price host functions

### PR DESCRIPTION
**Pull request by jakmeier**
https://github.com/near/NEPs/pull/372/

### CODE REVIEW

**Comment by jakmeier** [comment]
I believe the implementation is complete. Unit tests are okay, too.

But integration tests are just a rough draft so far. Maybe we need some refactoring to go in first. And also the snapshot testing doesn't work with receipt hashes, yet.

**Comment by jakmeier** [comment]
I think tests are mostly ready now, too.

I decided to also try snapshot tests here. It seems that any protocol upgrade could us a test of the form "this action produces an outcome like this before and like that afterwards".

In the case at hand, before the feature, there has to be an import error on the previously non-existing host function. Afterwards, the constructed contract must not fail (`SuccessValue(``)`) and the gas profile looks a certain way. So I was able to compare the gas profile manually and make sure it matches the specification. But the test itself only snapshots the outcome.

Probably we could refactor other tests in the file to follow a similar pattern.

File: integration-tests/src/tests/client/process_blocks.rs
Comment (akhi3030) [inline integration-tests/src/tests/client/process_blocks.rs:5205]: nit: should we rename some of the `*pessimistic*` names to `*purchased*` in the code?  The current scheme might be confusing in future.

File: runtime/near-vm-logic/src/context.rs
Comment (jakmeier) [reply runtime/near-vm-logic/src/context.rs:67]: Correct, no assignment other than initialisation. :) 
Check out the `VMContext {` construction inside `action.rs` of the diff.
```rust
        current_gas_price: apply_state.gas_price,
        receipt_gas_price: action_receipt.gas_price,
```

Comment (akhi3030) [reply runtime/near-vm-logic/src/context.rs:67]: Ah, missed that, thanks for the note!

File: runtime/runtime/src/actions.rs
Comment (akhi3030) [inline runtime/runtime/src/actions.rs:100]: nit: are we worried about someone swapping these around accidentally in the future as they have the same types?  In the past, I've tried the following two mechanisms to safeguard against such accidents:
- If one value is guaranteed to be an integer multiple of the other, store the first value as the original type and for the second we store the multipler.  I don't think this is applicable here.
- Introduce a `new` function that can add some checks e.g. if one value should be `>=` than the other, etc. and simply panic if invariants are broken.

No strong opinion about this, just thinking out loud.

**Review by akhi3030 — APPROVED** [review]
LGTM from my side.  I don't feel like I could fairly review all the tests though.  So maybe matklad or someone else should also take a look before we ship this.

**Comment by jakmeier** [comment]
> LGTM from my side. I don't feel like I could fairly review all the tests though. So maybe matklad or someone else should also take a look before we ship this. And I suppose we want to wait a bit more till all the discussions on the NEP are finalised.

Oh yeah let's definitively wait for the NEP discussion to come to a conclusion. Writing the code in parallel just helps to see what the changes mean in practice.

File: runtime/runtime/src/actions.rs
Comment (jakmeier) [reply runtime/runtime/src/actions.rs:100]: I don't think even `>=` is guaranteed to hold here. And IMO, accidentally swapping two field values must absolutely be detected by our test infrastructure, or else our testing strategy is not fit for the purpose. In this case, the integration tests would fail because the host functions return the wrong values.

File: chain/client/src/test_utils.rs
Comment (matklad) [inline chain/client/src/test_utils.rs:1598]: I'd maybe split this into a separate PR, there's a couple of questions here, like why we need to process epoch, or why do we pass in signer instead of account ID. Basically, if we want to bikeshed the good API here, let's do a separate PR. If we want to have something just for this PR, than this looks good, but let's not forget to revisit after merging -- suboptimal test helpers are a huge hazard.

File: runtime/near-test-contracts/contract-for-fuzzing-rs/src/lib.rs
Comment (matklad) [inline runtime/near-test-contracts/contract-for-fuzzing-rs/src/lib.rs:31]: To bikeshed mode, `block_gas_price` perhaps?

File: runtime/near-vm-logic/src/context.rs
Comment (matklad) [inline runtime/near-vm-logic/src/context.rs:67]: Let's use he same name for this field and the host fn

File: integration-tests/src/tests/client/process_blocks.rs
Comment (matklad) [inline integration-tests/src/tests/client/process_blocks.rs:5228]: Let's use 

https://github.com/near/nearcore/blob/27d2e5acb67698c59a1bd5e7bbac1234f062195d/runtime/near-test-contracts/src/lib.rs#L9

here to abstract away which wasm parser we use

File: integration-tests/src/tests/client/process_blocks.rs
Comment (matklad) [inline integration-tests/src/tests/client/process_blocks.rs:5205]: I am wondering if we should more explicitly set various configs that affect prices in tests? 

That's tradeoff between:

* a test explicitly setting its inputs to avoid unrelated changes
* a test checking not only the code, but also the actual configuration we use

File: integration-tests/src/tests/client/snapshots/integration_tests__tests__client__process_blocks__runtime_gas_price__100Tgas_purchased_gas_price before GasPriceHostFn.snap
Comment (matklad) [inline integration-tests/src/tests/client/snapshots/integration_tests__tests__client__process_blocks__runtime_gas_price__100Tgas_purchased_gas_price before GasPriceHostFn.snap:33]: If we expect to use such tests frequently, we probably should come up with more compact and readable representation for execution outcome.

**Review by matklad — COMMENTED** [review]
Broad :+1: from me

File: chain/client/src/test_utils.rs
Comment (jakmeier) [reply chain/client/src/test_utils.rs:1598]: Yeah, let's refactor a bit in a separate PR. Also because this PR here depends on NEP decisions and I want a better test infrastructure asap, to also write test for other implementations.

Created #7217 with a first suggestion, please bikeshed on that PR :)

File: integration-tests/src/tests/client/process_blocks.rs
Comment (jakmeier) [reply integration-tests/src/tests/client/process_blocks.rs:5078]: I think it makes sense to specify the gas related parameters, ideally these tests shouldn't fail when we update those values. So I did this:
```
genesis.config.protocol_version = old_protocol_version;
genesis.config.min_gas_price = 10u128.pow(9);
```

**Comment by jakmeier** [comment]
NEP discussion on this is still outstanding, will mark this as a draft as there is nothing more to review until we have a conclusion there.

**Comment by chefsale** [comment]
Is this PR still alive? 🙃

**Comment by jakmeier** [comment]
Almost dead, or at least in deep sleep for now. :)

Not enough interest from the community has been shown. So we dropped the ball on this and focus on other things.

There is also reason to wait. Paying for others using access keys was one use case here. But this can also be addressed with meta transactions. So we also want to wait and see how that will end up being implemented.

Personally, I think this could still be valuable. Using prepaid access keys have shorter latency than meta transaction. And link drops are harder to pull of with meta transactions compared to access keys. And I would imagine some users also want to stop doing certain actions if the gas price is too high. 

But it's up to smart contract devs that need to raise their voice if they want this feature.

**Comment by matklad** [comment]
As a procedural note, I think the right venue to discuss this at this point is the NEP PR: https://github.com/near/NEPs/pull/372, as this is still blocked on NEP.

NEP discussions tend to spill over to all various kind of places, so it usually is helpful to try to keep them centralized :)

**Pull request closed without merging**
